### ISSUE TITLE
Walrus comprehension rebind checking behavior

### ISSUE THREAD

**Issue by 799a973d-479e-458f-b448-ca9b830a935e** [python/cpython#87447]
BPO | [43281](https://bugs.python.org/issue43281)
--- | :---
Nosy | @gvanrossum, @emilyemorehouse, @gousaiyang

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-02-21.01:20:04.826>
labels = ['type-bug', '3.8', '3.9', '3.10']
title = 'Walrus comprehension rebind checking behavior'
updated_at = <Date 2021-02-27.05:46:20.917>
user = 'https://github.com/gousaiyang'
```

bugs.python.org fields:
```python
activity = <Date 2021-02-27.05:46:20.917>
actor = 'gvanrossum'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2021-02-21.01:20:04.826>
creator = 'gousaiyang'
dependencies = []
files = []
hgrepos = []
issue_num = 43281
keywords = []
message_count = 2.0
messages = ['387433', '387768']
nosy_count = 3.0
nosy_names = ['gvanrossum', 'emilyemorehouse', 'gousaiyang']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue43281'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

### Linked PRs
* gh-100581

**Comment by sobolevn** [comment]
I had some free time to learn more about `symtable.c` and fix this :) 
So, I hope my PR is in the right direction!

**Comment by hauntsaninja** [comment]
I think it probably needs a little more discussion whether this should be fixed.

I found it useful to look at a slightly simpler example than the test case, like:
```
index = 0
a = [0, 0, 0]
[(index := i) for i, a[index] in enumerate([9,8,7])]
```

Arguments in favour of changing:
- Like OP points out, the rationale in PEP 572 for outlawing these only applies to `a` and `b`; the scoping is clear for the other variables (so we'd have to invent new reasons to outlaw them)
- I think it's probably a little surprising that one gets `SyntaxError: assignment expression cannot rebind comprehension iteration variable 'index'` from the above
- It's seemingly a small change to fix
- I could squint my eyes to believing there's some use case for it

Arguments against changing:
- Any comprehension that does this is going to be at least a little cursed. No concrete use case has been proposed
- Python 3.8 has been around for a couple years and this hasn't really come up
- There are stricter-than-strictly-necessary restrictions on assignment expressions in comprehensions that we're okay with, that I find more surprising than this
- Changing this could necessitate changes in other Python implementations or linters
- It's maybe a little weird that you'd be able to do `[(a := [i]) for i, a[0] in enumerate([9,8,7])]` but not be able to do `[(a := i) for i, a in enumerate([9,8,7])]`

Overall, I think I lean against, but I could be convinced otherwise!

**Comment by sobolevn** [comment]
My main argument to fix this is that error message does not show the reality:

```
>>> [(h := 1) for a, (*b, c[d+e::f(g)], h.i) in [1]]
  File "<stdin>", line 1
SyntaxError: assignment expression cannot rebind comprehension iteration variable 'h'
```

But, `h` is not a comprehension iteration variable, `h.i` is.

So, in my opinion it is clearly a bug. But, there are two ways of fixing it:
1. Allow `h.i`, as in my PR
2. Try changing the error message to something like `assignment expression cannot rebind variable 'h' that is used in comprehension iteration expression` (or whatever)

**Comment by gvanrossum** [comment]
I feel for the OP and I agree with several arguments in favor of this fix, *given that it's a simple fix*. Shantanu's arguments against feel pretty weak (probably intentionally so: "at least a little", "maybe a little").

Maybe the other thing (disallowing named expressions in the iterable altogether) can be fixed in another PR?

**Comment by sobolevn** [comment]
> Maybe the other thing (disallowing named expressions in the iterable altogether) can be fixed in another PR?

Are you talking about forbidding `:=` in `[(h := 1) for a` part? If I understand you correctly, this is would be a breaking change, because even PEP 572 references this use-case: https://peps.python.org/pep-0572/#scope-of-the-target

```python
# Compute partial sums in a list comprehension
total = 0
partial_sums = [total := total + v for v in values]
print("Total:", total)
```

However, I probably just didn't get your idea :)

**Comment by gvanrossum** [comment]
I was talking about this from Shantanu:

> * There are stricter-than-strictly-necessary restrictions on assignment expressions in comprehensions that we're okay with, that I find more surprising than this (disallowing them entirely in the iterable)

This is about things like:
```
>>> [x for x in (a := range(3))]
  File "<stdin>", line 1
SyntaxError: assignment expression cannot be used in a comprehension iterable expression
>>> 
```

**Comment by emilyemorehouse** [comment]
Fixed in #100581
